gobblin-runtime/src/main/java/org/apache/gobblin/runtime/mapreduce/MRJobLauncher.java

```diff
@@ -389,23 +391,61 @@ static void serializeJobState(FileSystem fs, Path mrJobDir, Configuration conf,
   private void addJars(Path jarFileDir, String jarFileList, Configuration conf)
       throws IOException {
     LocalFileSystem lfs = FileSystem.getLocal(conf);
+    Set<FileStatus> toRetryFileStatus = new HashSet<>();
+    int retryCount = 0 ;
     for (String jarFile : SPLITTER.split(jarFileList)) {
       Path srcJarFile = new Path(jarFile);
       FileStatus[] fileStatusList = lfs.globStatus(srcJarFile);
       for (FileStatus status : fileStatusList) {
-        // SNAPSHOT jars should not be shared, as different jobs may be using different versions of it
-        Path baseDir = status.getPath().getName().contains("SNAPSHOT") ? this.unsharedJarsDir : jarFileDir;
-        // DistributedCache requires absolute path, so we need to use makeQualified.
-        Path destJarFile = new Path(this.fs.makeQualified(baseDir), status.getPath().getName());
+        Path destJarFile = calculateDestJarFile(status, jarFileDir);
+        // Adding destJarFile into HDFS until it exists
         if (!this.fs.exists(destJarFile)) {
           // Copy the jar file from local file system to HDFS
-          this.fs.copyFromLocalFile(status.getPath(), destJarFile);
+          // Set the first parameter as false for not deleting sourceFile
+          // Set the second parameter as false for not overwriting existing file on the target, by default it is true.
+            // If the file is preExisted but overwrite flag set to false, then an IOException if thrown.
+          try {
```

Comment:
You can avoid duplicating the code by adding the retry here. Any reason the retry is done later?
We should add a check for file size to make sure that the file is complete before we continue.

Reply:
I am trying to increase the window time between a failure of copying and a retry for a specific file. I can refactor it if it is not necessary in this case. 

By "add a check for file size", does that mean we are not only checking the file existence but compare the file size when we decided that copy is not necessary for a file?  Does this check mainly to avoid jar file being corrupted ? It doesn't seem to be necessary since we have disable the overwrite flag.

Reply:
The failure is most likely due to the file existing, so I think an immediate retry should succeed. The file size check is to make sure that the file copied by the other process is complete before proceeding to use it in this process. The file can exist, but still be in the process of getting filled in.

Reply:
Got it.  Thanks for explanation.